### PR TITLE
[Feature] Named dims

### PR DESCRIPTION
**Pull request by vmoens**
## Description

Introduces optional named dimensions

Helps with features such as https://github.com/pytorch/rl/pull/1080

Covers
- [x] indexing
- [x] cat
- [x] stack
- [x] squeeze
- [x] unsqueeze
- [x] clone
- [x] permute
- [x] expand
- [x] sub-tds
- [x] flatten
- [x] apply
- [x] gather
- [x] masked_fill
- [x] select
- [x] memmap_like
- [x] detach
- [x] unbind
- [x] split
- [x] any / all with dim
- [x] set and set_

Reshape and view loose the names as expected

TODO
- [ ] transpose => not implemented yet
- [ ] scatter => not implemented yet
- [ ] where => not implemented yet

### CODE REVIEW

**Comment by vmoens** [comment]
> ## ⚠
> ⚠
> Warning
>  Result of CPU Benchmark Tests
> ### Total Benchmarks: 46. Improved: 
> 14
> . Worsened: 
> 2
> .

Seriously?

File: tensordict/tensordict.py [hunk 1]
```diff
@@ -360,6 +361,132 @@ def batch_size(self) -> torch.Size:
...
+        greedily; it is expanded in-place to fill names to the same length as
+        self.dim() using names from the corresponding indices of self.names.
+
+        Returns: the tensordict with dimensions named accordingly.
+
+        """
+        # replace ellipsis if any
+        names_copy = copy(names)
+        if any(name is Ellipsis for name in names):
+            ellipsis_name = [NO_DEFAULT for _ in self.ndim - len(names)]
```
Comment (tcbegley) [inline tensordict/tensordict.py:412]: This doesn't look right. At the least it's missing `range`, but also I think there's an off by one error (need to correct for the fact that we're removing `Ellipsis` which is counted in `len(names)`)

File: tensordict/tensordict.py [hunk 2]
```diff
@@ -360,6 +361,132 @@ def batch_size(self) -> torch.Size:
...
+        self._rename_subtds(self.names)
+        return self
+
+    def rename(self, *names, **rename_map):
+        clone = self.clone(recurse=False)
+        if len(names) == 1 and names[0] is None:
+            clone.names = None
+        if rename_map and names:
+            raise ValueError(
+                "Passed both a name map and a name list. " "Only one is accepted."
```
Comment (tcbegley) [inline tensordict/tensordict.py:443]:
```suggestion
                "Passed both a name map and a name list. Only one is accepted."
```

**Review by tcbegley — COMMENTED** [review]
This looks like a good start.

I think we should add a test for `refine_names` as there's no coverage there atm.

Also should names be propagated to tensors in the tensor dict? I'm seeing for example

```python
import torch
from tensordict import TensorDict

td = TensorDict({"a": torch.rand(10)}, [10], names=["my_dim"])
assert td.names == ["my_dim"]
assert td["a"].names == (None,)
```

File: tensordict/tensordict.py [hunk 3]
```diff
@@ -360,6 +361,132 @@ def batch_size(self) -> torch.Size:
...
+        # replace ellipsis if any
+        names_copy = copy(names)
+        if any(name is Ellipsis for name in names):
+            ellipsis_name = [NO_DEFAULT for _ in self.ndim - len(names)]
+            names = []
+            for name in names_copy:
+                if name is Ellipsis:
+                    names += ellipsis_name
+                else:
+                    names.append(name)
```
[the remark this answers is not in the document]
Comment (vmoens) [reply tensordict/tensordict.py:418]: Can you have more than one ellipsis?
[..., 1, ...] will not work
[1, ..., ...] is dubious
[..., ..., 1] also
[..., 1, ..., 1] 
others?

[the remark this answers is not in the document]
Comment (tcbegley) [reply tensordict/tensordict.py:418]: I don't think there would be a legitimate usage, so I guess maybe fine to leave it on the user to provide reasonable input rather than to do additional runtime checks that might slow things down 👍

File: tensordict/tensordict.py (hunk 1 above)
Comment (vmoens) [reply tensordict/tensordict.py]: Edited!
